pkg/peertls/extensions/extensions.go

```diff
@@ -0,0 +1,211 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package extensions
+
+import (
+	"crypto"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/pkg/peertls"
+	"storj.io/storj/pkg/pkcrypto"
+)
+
+const (
+	// RevocationBucket is the bolt bucket to store revocation data in
+	RevocationBucket = "revocations"
+)
+
+var (
+	// AllHandlers holds all registered extension handlers
+	AllHandlers HandlerFactories
+
+	// CAWhitelistSignedLeafHandler verifies that the leaf cert of the remote peer's
+	// identity was signed by one of the CA certs in the whitelist.
+	CAWhitelistSignedLeafHandler = NewHandler(
+		&SignedCertExtID, caWhitelistSignedLeafHandler,
+	)
+
+	// NB: 2.999.X is reserved for "example" OIDs
+	// (see http://oid-info.com/get/2.999)
+	// 2.999.1.X -- storj general/misc. extensions
+
+	// SignedCertExtID is the asn1 object ID for a pkix extensionHandler holding a
+	// signature of the cert it's extending, signed by some CA (e.g. the root cert chain).
+	// This extensionHandler allows for an additional signature per certificate.
+	SignedCertExtID = ExtensionID{2, 999, 1, 1}
+	// RevocationExtID is the asn1 object ID for a pkix extensionHandler containing the
+	// most recent certificate revocation data
+	// for the current TLS cert chain.
+	RevocationExtID = ExtensionID{2, 999, 1, 2}
+
+	// Error is used when an error occurs while processing an extension.
+	Error = errs.Class("extension error")
+
+	// ErrVerifyCASignedLeaf is used when a signed leaf extension signature wasn't produced
+	// by any CA in the whitelist.
+	ErrVerifyCASignedLeaf = Error.New("leaf not signed by any CA in the whitelist")
+	// ErrUniqueExtensions is used when multiple extensions have the same Id
+	ErrUniqueExtensions = Error.New("extensions are not unique")
+)
+
+// ExtensionID is an alias to an `asn1.ObjectIdentifier`.
+type ExtensionID asn1.ObjectIdentifier
+
+// Config is used to bind cli flags for determining which extensions will
+// be used by the server
+type Config struct {
+	Revocation          bool `default:"true" help:"if true, client leaves may contain the most recent certificate revocation for the current certificate"`
+	WhitelistSignedLeaf bool `default:"false" help:"if true, client leaves must contain a valid \"signed certificate extension\" (NB: verified against certs in the peer ca whitelist; i.e. if true, a whitelist must be provided)"`
+}
+
+// Options holds common options for use in handling extensions.
+type Options struct {
+	PeerCAWhitelist []*x509.Certificate
+	RevDB           RevocationDB
+}
+
+// HandlerFactories is a collection of `HandlerFactory`s for convenience.
+// Defines `Register` and `WithOptions` methods.
+type HandlerFactories []*HandlerFactory
```

Comment:
These factories etc. make me feel like there is a nicer design around the corner, but I'm unsure what it exactly is.

Reply:
@egonelbre Are you of the opinion that factory methods are an anti pattern ?

Reply:
i'm interested in the answer as well

Reply:
maybe just more meaningful names would already be a good start